### PR TITLE
focal loss 추가 + 클래스별 리콜와 프리시전

### PR DESCRIPTION
**Pull request by ParkJinHyeock**
각 class별로 recall과 precision을 계산해서 텐서보드에 저장하는 것과
loss를 focal loss와 bce중에 선택하는 코드를 추가하였습니다. config를 이용해 알파와 gamma를 조절할 수 있습니다.

### CODE REVIEW

File: metrics.py [hunk 1]
```diff
@@ -87,14 +96,20 @@ def update_block_states(self, y_true_block, y_pred_block):
...
+        self.class_tn += tf.reduce_sum(true_negative, axis=(-3,-2))
+        self.class_tp += tf.reduce_sum(true_positives, axis=(-3,-2))
+        
         self.FN += tf.math.reduce_sum(false_negative)
         self.FP += tf.math.reduce_sum(false_positive)
         loc_FN = tf.math.reduce_sum(false_negative, axis=(-2, -1))
         loc_FP = tf.math.reduce_sum(false_positive, axis=(-2, -1))
 
         ''' when a class exists in both y_true and y_pred '''
-        true_positives = true_classes * pred_classes
```
Comment (daniel03c1) [inline metrics.py:97]: 이 줄 빼셨는데 빼도 잘 돌아가나요?

File: train.py [hunk 2]
```diff
@@ -90,6 +92,40 @@ def iterloop(model, dataset, sed_loss, doa_loss, metric_class, config, epoch, wr
...
+    'female scream': recall[5].numpy(),
+    'femael speech': recall[6].numpy(),
+    'burning fire': recall[7].numpy(),
+    'footsteps': recall[8].numpy(),
+    'knocking on door': recall[9].numpy(),
+    'man scream': recall[10].numpy(),
+    'man speech': recall[11].numpy(),
+    'ringing phone': recall[12].numpy(),
+    'piano': recall[13].numpy(),
+    }, epoch)
```
Comment (daniel03c1) [inline train.py:110]: 코드 스타일 관련인데요 'alarm'부터 'piano', }, epoch)까지 다 writer.add_scalars와 같은 선상에 놓여 있는데,
이러면 함수의 인자들이니까 적어도 한 탭 해주면 좋을 것 같습니다.

File: losses.py [hunk 3]
```diff
@@ -12,3 +12,17 @@ def MMSE(y_true, y_pred):
...
+        self.alpha = alpha
+        self.gamma = gamma
+    
+    def __call__(self, y_true, y_pred):
+        eps = 1e-7
+        y_pred = tf.clip_by_value(y_pred, eps, 1-eps)
+        focal = - y_true * self.alpha * tf.pow(1 - y_pred, self.gamma) * tf.math.log(y_pred)\
+                - (1-y_true) * self.alpha * tf.pow(y_pred, self.gamma) * tf.math.log(1-y_pred)
+
+        return tf.reduce_mean(focal)
```
Comment (lsg1213) [inline losses.py:28]: 끝에 1줄 넣어주시면 더 좋을 것 같습니다

File: metrics.py [hunk 4]
```diff
@@ -25,6 +25,10 @@ def reset_states(self):
 
         self.total_DE = tf.zeros([], tf.float32)
         self.DE_TP = tf.zeros([], tf.float32)
+        self.class_tp = tf.zeros([14], tf.float32)
+        self.class_fp = tf.zeros([14], tf.float32)
+        self.class_tn = tf.zeros([14], tf.float32)
+        self.class_fn = tf.zeros([14], tf.float32)
```
Comment (lsg1213) [inline metrics.py:31]: class를 dynamic하게 받으면 어떨까요? 2021 가서는 class 개수가 바뀌어서요

Comment (ParkJinHyeock) [reply metrics.py:31]: 아 네 알겠습니다. num classes를 넣어주는 방향으로 수정해야할 것 같네요

**Comment by lsg1213** [comment]
mixing은 뭔지 알 수 있을까요?

File: data_loader.py [hunk 5]
```diff
@@ -40,7 +40,7 @@ def apply_ops(dataset, operations):
         return dataset
 
     dataset = apply_ops(dataset, preprocessing)
-    dataset = dataset.cache()
+    # dataset = dataset.cache()
```
Comment (daniel03c1) [inline data_loader.py:43]: cache를 왜 지우신건가요?
